rusk-wallet/src/lib/cache.rs

```diff
@@ -0,0 +1,86 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+
+use dusk_bytes::{DeserializableSlice, Serializable};
+use phoenix_core::Note;
+use rusqlite::{params, Connection};
+
+const TABLE_NOTES: &str =
+    "CREATE TABLE if not exists notes (note BLOB, spendkey BLOB)";
+const TABLE_CACHE: &str =
+    "CREATE TABLE if not exists cache (block BIGINT, spendkey BLOB)";
+
+const QUERY_BLOCKHEIGHT: &str = "SELECT block FROM cache WHERE spendkey = ?";
+const UPDATE_BLOCKHEIGHT: &str = "update cache set block=? WHERE spendkey = ?";
+const INSERT_BLOCKHEIGHT: &str =
+    "INSERT INTO cache (block, spendkey) VALUES (?1, ?2)";
+    
+const INSERT_NOTES: &str = "INSERT INTO notes (note, spendkey) values (?1, ?2)";
+const QUERY_NOTES: &str = "SELECT note FROM notes WHERE spendkey = ?";
+
+pub(crate) fn open(data_dir: &PathBuf) -> Connection {
+    let db_path = Path::new(data_dir).join("cache.db");
+
+    let cache = Connection::open(db_path).unwrap();
+    cache.execute(TABLE_NOTES, []).unwrap();
+    cache.execute(TABLE_CACHE, []).unwrap();
+    cache
+}
+
+pub(crate) fn last_blockheight(conn: &Connection, spendkey: &[u8]) -> u64 {
+    let cached_blockheight =
+        conn.query_row_and_then(QUERY_BLOCKHEIGHT, [spendkey], |row| {
+            row.get(0)
+        });
+    cached_blockheight.unwrap_or(0u64)
+}
+
+pub(crate) fn persist_blockheight(
+    conn: &Connection,
+    spendkey: &[u8],
+    actual_block: u64,
+) {
+    if conn
+        .execute(UPDATE_BLOCKHEIGHT, params!(actual_block, spendkey))
+        .unwrap()
+        == 0
+    {
+        conn.execute(INSERT_BLOCKHEIGHT, params!(actual_block, spendkey))
+            .unwrap();
+    }
+}
+
+pub(crate) fn persist_notes(
+    conn: &Connection,
+    spendkey: &[u8],
+    notes: &[Note],
+) {
+    notes.iter().for_each(|&n| {
+        let mut insert_stats = conn.prepare(INSERT_NOTES).unwrap();
+        insert_stats
+            .execute([n.to_bytes().to_vec(), spendkey.to_vec()])
+            .unwrap();
+    });
+}
+
+pub(crate) fn cached_notes(
+    conn: &Connection,
+    spendkey: &[u8],
+) -> HashMap<Vec<u8>, Note> {
+    let mut stmt = conn.prepare(QUERY_NOTES).unwrap();
+    let person_iter = stmt
+        .query_map([spendkey], |row| {
+            let note_bytes: Vec<u8> = row.get(0).unwrap();
+            let note = Note::from_slice(&note_bytes[..]).unwrap();
+            Ok((note.hash().to_bytes().to_vec(), note))
+        })
+        .unwrap();
```

Comment:
Once again, a proper `Error` here should be introduced to the crate and passed up to the user.